phases/11-llm-engineering/09-function-calling/code/function_calling.py

```
@@ -100,9 +100,14 @@ def read_file(path):
     if language != "python":
         return {"error": True, "message": f"Language '{language}' not supported. Only 'python' is available."}
-    forbidden = ["import os", "import sys", "import subprocess", "exec(", "eval(", "__import__", "open("]
+    # TOY SANDBOX: This blocklist is for demonstration only. Do not use in production.
+    # Real-world sandboxing requires Docker, gVisor, or dedicated runtimes (e.g. E2B).
+    forbidden = [
+        r"\b(?:import|from)\s+(?:os|sys|subprocess|shlex|pty|ctypes)\b",
+        r"\b(exec|eval|__import__|open|getattr|setattr|delattr|hasattr)\s*\(",
+    ]
     for pattern in forbidden:
-        if pattern in code:
+        if re.search(pattern, code, re.IGNORECASE | re.MULTILINE):
```

Ruff (0.15.15)

[error] Undefined name `re`

(F821)